inc/bucket.php

```diff
@@ -0,0 +1,135 @@
+<?php
+/**
+ * Each page has 50,000 / CORE_SITEMAPS_POSTS_PER_BUCKET buckets.
+ */
+
+defined( 'ABSPATH' ) || die();
+
+/**
+ * Register the Sitemap Bucket custom post-type.
+ */
+function core_sitemaps_bucket_register() {
+	$labels = array(
+		'name'          => _x( 'Sitemap Buckets', 'Sitemap Bucket General Name', 'core-sitemaps' ),
+		'singular_name' => _x( 'Sitemap Bucket', 'Sitemap Bucket Singular Name', 'core-sitemaps' ),
+	);
+	$args   = array(
+		'label'           => __( 'Sitemap Bucket', 'core-sitemaps' ),
+		'description'     => __( 'Bucket of sitemap links', 'core-sitemaps' ),
+		'labels'          => $labels,
+		'supports'        => array( 'editor', 'custom-fields' ),
+		'can_export'      => false,
+		'rewrite'         => false,
+		'capability_type' => 'post',
+	);
+	register_post_type( CORE_SITEMAPS_CPT_BUCKET, $args );
+}
+
+/**
+ * Calculate the sitemap bucket number the post belongs to.
+ *
+ * @param int $post_id Post ID.
+ *
+ * @return int Sitemap Page pagination number.
+ */
+function core_sitemaps_page_calculate_bucket_num( $post_id ) {
+	// TODO this lookup might need to be more refined and set min/max
+	return 1 + (int) floor( $post_id / CORE_SITEMAPS_POSTS_PER_BUCKET );
+}
+
+/**
+ * Get the Sitemap Page for a pagination number.
+ *
+ * @param string $post_type Registered post-type.
+ * @param int    $start_bucket Sitemap Page pagination number.
+ *
+ * @param int    $max_buckets Number of buckets to return.
+ *
+ * @return bool|int[]|WP_Post[] Zero or more Post objects of the type CORE_SITEMAPS_CPT_PAGE.
+ */
+function core_sitemaps_bucket_lookup( $post_type, $start_bucket, $max_buckets = 1 ) {
+	$page_query            = new WP_Query();
+	$registered_post_types = core_sitemaps_registered_post_types();
+	if ( false === isset( $registered_post_types[ $post_type ] ) ) {
+		return false;
+	}
+	$bucket_meta = array(
+		array(
+			'key'   => 'post_type',
+			'value' => $post_type,
+		),
+	);
+	if ( 1 === $max_buckets ) {
+		// One bucket.
+		$bucket_meta[] = array(
+			'key'   => 'bucket_num',
+			'value' => $start_bucket,
+		);
+	} else {
+		// Range query.
+		$bucket_meta[] = array(
+			'key'     => 'bucket_num',
+			'value'   => array( $start_bucket, $start_bucket + $max_buckets - 1 ),
+			'type'    => 'numeric',
+			'compare' => 'BETWEEN',
+		);
+	}
+
+	$query_result = $page_query->query(
```

Comment:
We'll want to avoid meta queries that are based on meta values, because they are generally slow. We'll probably need to get all `post_meta` rows that use the key of our buckets and loop through each of them to find the one that the post ID we're looking for is in and then map that to the bucket (CPT) ID.